From 1fc4e6da7bf4081fb89c9f40b33ac2b422f437c7 Mon Sep 17 00:00:00 2001
From: Angela Fong <angela.fong@joyent.com>
Date: Sun, 22 Oct 2017 19:55:43 -0700
Subject: [PATCH] ADMINUI-2370 Disk provisioned is misleading as it includes
 the reserved amount Reviewed by: Marsell Kukuljevic <marsell@joyent.com>
 Approved by: Marsell Kukuljevic <marsell@joyent.com>

---
 less/server.less                              |  8 ++-
 .../components/pages/server/disk-overview.jsx | 52 ++++++++++++++++---
 .../pages/server/disk-utilization-circle.jsx  |  7 ++-
 .../pages/server/memory-overview.jsx          | 35 +++++++++++--
 4 files changed, 88 insertions(+), 14 deletions(-)

diff --git a/less/server.less b/less/server.less
index 97057294..7779333a 100644
--- a/less/server.less
+++ b/less/server.less
@@ -22,6 +22,7 @@
   .provisionable-disk,
   .provisioned-disk,
   .total-disk,
+  .unreserved-disk,
   .reserved-disk {
     .make-sm-column(2, 0);
 
@@ -31,7 +32,7 @@
     .title { text-transform: uppercase; color: #666;  font-size: 12px; }
   }
 
-  .total-disk { opacity: 0.5; }
+  .unreserved-disk, .total-disk { opacity: 0.5; }
 
   .memory-overview h5 {
     text-transform: uppercase;
@@ -67,6 +68,7 @@
 
   .provisionable-disk .value,
   .provisioned-disk .value,
+  .reserved-disk .value,
   .provisionable-memory .value,
   .provisioned-memory .value,
   .reserved-memory .value {
@@ -77,14 +79,18 @@
   .provisioned-disk .value,
   .provisionable-memory .value,
   .provisioned-memory .value,
+  .reserved-disk .value,
   .reserved-memory .value,
   .total-disk .value,
+  .unreserved-disk .value,
   .unreserved-memory .value,
   .total-memory .value {
     color: #888;
   }
 
   .unreserved-memory, .total-memory { opacity: 0.5; }
+
+  .unreserved-disk,
   .unreserved-memory {
     border-left: 1px solid #aaa;
   }
diff --git a/www/js/components/pages/server/disk-overview.jsx b/www/js/components/pages/server/disk-overview.jsx
index 8acd8ac5..5159617a 100644
--- a/www/js/components/pages/server/disk-overview.jsx
+++ b/www/js/components/pages/server/disk-overview.jsx
@@ -12,6 +12,7 @@ var React = require('react');
 var BackboneMixin = require('../../_backbone-mixin');
 var ServerDiskUtilizationCircle = require('./disk-utilization-circle');
 var utils = require('../../../lib/utils');
+var POOL_USABLE_RATIO = 0.94;
 
 var ServerDiskOverview = React.createClass({
     mixins: [BackboneMixin],
@@ -24,10 +25,15 @@ var ServerDiskOverview = React.createClass({
             return null;
         }
 
-        var provisioned = server.disk_pool_size_bytes - (server.unreserved_disk * 1048576);
+//      DAPI applies a pool usage ratio in its calculation of provisionable disk.
+//      Unfortunately that's named unreserved_disk which is different from what it means
+//      in this context. Here "unreserved" means what is not reserved by the system.
+        var unreserved = server.disk_pool_size_bytes * POOL_USABLE_RATIO;
+        var reserved = server.disk_pool_size_bytes - unreserved;
         var provisionable = server.unreserved_disk * 1048576;
+        var provisioned = unreserved - provisionable;
         var total = server.disk_pool_size_bytes;
-        
+
         if (provisioned < 0 || !provisioned) { provisioned = 0; }
         if (provisionable < 0 || !provisionable) {
             if (!provisionable && server.disk_pool_size_bytes) { provisionable = server.disk_pool_size_bytes; }
@@ -37,6 +43,8 @@ var ServerDiskOverview = React.createClass({
 
         var provisioned = utils.getReadableSize(provisioned);
         var provisionable = utils.getReadableSize(provisionable);
+        var reserved = utils.getReadableSize(reserved);
+        var unreserved = utils.getReadableSize(unreserved);
         var total = utils.getReadableSize(server.disk_pool_size_bytes);
 
         return <div className="disk-overview">
@@ -51,16 +59,48 @@ var ServerDiskOverview = React.createClass({
                 </div>
                 <div className="provisionable-disk">
                     <div className="value">{provisionable.value + ' ' + provisionable.measure}</div>
-                    <div className="title">Provisionable</div>
+                    <div className="title">
+                        <a data-toggle="tooltip" data-placement="bottom" data-trigger="hover"
+                            title="Amount of disk currently available for provisioning.">
+                            Provisionable
+                        </a>
+                    </div>
                 </div>
                 <div className="provisioned-disk">
                     <div className="value">{provisioned.value + ' ' + provisioned.measure}</div>
-                    <div className="title">Provisioned</div>
+                    <div className="title">
+                        <a data-toggle="tooltip" data-placement="bottom" data-trigger="hover"
+                            title="Amount of disk already committed to provisioned instances.">
+                            Provisioned
+                        </a>
+                    </div>
+                </div>
+                <div className="reserved-disk">
+                    <div className="value">{reserved.value + ' ' + reserved.measure}</div>
+                    <div className="title">
+                        <a data-toggle="tooltip" data-placement="bottom" data-trigger="hover"
+                            title="Amount of disk reserved for system use.">
+                            Reserved
+                        </a>
+                    </div>
+                </div>
+                <div className="unreserved-disk">
+                    <div className="value">{unreserved.value + ' ' + unreserved.measure}</div>
+                    <div className="title">
+                        <a data-toggle="tooltip" data-placement="bottom" data-trigger="hover"
+                            title="Total pool size minus disk reserved for system use.">
+                            Unreserved
+                        </a>
+                    </div>
                 </div>
-
                 <div className="total-disk">
                     <div className="value">{total.value + ' ' + total.measure}</div>
-                    <div className="title">Total</div>
+                    <div className="title">
+                        <a data-toggle="tooltip" data-placement="bottom" data-trigger="hover"
+                            title="Total disk pool size.">
+                            Total
+                        </a>
+                    </div>
                 </div>
             </div>
         </div>;
diff --git a/www/js/components/pages/server/disk-utilization-circle.jsx b/www/js/components/pages/server/disk-utilization-circle.jsx
index 46358875..9709025b 100644
--- a/www/js/components/pages/server/disk-utilization-circle.jsx
+++ b/www/js/components/pages/server/disk-utilization-circle.jsx
@@ -12,6 +12,7 @@ var $ = require('jquery');
 require('epoch');
 var React = require('react');
 var BackboneMixin = require('../../_backbone-mixin');
+var POOL_USABLE_RATIO = 0.94;
 
 var ServerMemoryUtilizationCircle = React.createClass({
     mixins: [BackboneMixin],
@@ -28,8 +29,9 @@ var ServerMemoryUtilizationCircle = React.createClass({
     getChartData: function() {
         var server = this.props.server.toJSON();
 
-        var provisioned = server.disk_pool_size_bytes - (server.unreserved_disk * 1048576);
+        var reserved = server.disk_pool_size_bytes * (1 - POOL_USABLE_RATIO);
         var provisionable = server.unreserved_disk * 1048576;
+        var provisioned = server.disk_pool_size_bytes - reserved - provisionable;
 
         if (provisioned < 0 || !provisioned) { provisioned = 0; }
         if (provisionable < 0 || !provisionable) {
@@ -38,6 +40,7 @@ var ServerMemoryUtilizationCircle = React.createClass({
         }
 
         var pieData = [
+            {label: 'Reserved', value: reserved },
             {label: 'Provisioned', value: provisioned },
             {label: 'Provisionable', value: provisionable },
         ];
@@ -72,7 +75,7 @@ var ServerMemoryUtilizationCircle = React.createClass({
         var percentmTop = (-(diameter/2) - 9) + 'px';
         var server = this.props.server.toJSON();
 
-        var provisioned = server.disk_pool_size_bytes - (server.unreserved_disk * 1048576);
+        var provisioned = server.disk_pool_size_bytes * POOL_USABLE_RATIO - (server.unreserved_disk * 1048576);
         var total = server.disk_pool_size_bytes;
 
         if (provisioned < 0 || !provisioned) { provisioned = 0; }
diff --git a/www/js/components/pages/server/memory-overview.jsx b/www/js/components/pages/server/memory-overview.jsx
index eff86e0b..7626059c 100644
--- a/www/js/components/pages/server/memory-overview.jsx
+++ b/www/js/components/pages/server/memory-overview.jsx
@@ -40,23 +40,48 @@ var ServerMemoryOverview = React.createClass({
                 </div>
                 <div className="provisionable-memory">
                     <div className="value">{provisionable.value + ' ' + provisionable.measure}</div>
-                    <div className="title">Provisionable</div>
+                    <div className="title">
+                        <a data-toggle="tooltip" data-placement="bottom" data-trigger="hover"
+                            title="Amount of memory currently available for provisioning.">
+                            Provisionable
+                        </a>
+                    </div>
                 </div>
                 <div className="provisioned-memory">
                     <div className="value">{provisioned.value + ' ' + provisioned.measure}</div>
-                    <div className="title">Provisioned</div>
+                    <div className="title">
+                        <a data-toggle="tooltip" data-placement="bottom" data-trigger="hover"
+                            title="Amount of memory already committed to provisioned instances.">
+                            Provisioned
+                        </a>
+                    </div>
                 </div>
                 <div className="reserved-memory">
                     <div className="value">{reserved.value + ' ' + reserved.measure}</div>
-                    <div className="title">Reserved</div>
+                    <div className="title">
+                        <a data-toggle="tooltip" data-placement="bottom" data-trigger="hover"
+                            title="Amount of memory reserved for system use.">
+                            Reserved
+                        </a>
+                    </div>
                 </div>
                 <div className="unreserved-memory">
                     <div className="value">{unreserved.value + ' ' + unreserved.measure}</div>
-                    <div className="title">Unreserved</div>
+                    <div className="title">
+                        <a data-toggle="tooltip" data-placement="bottom" data-trigger="hover"
+                            title="Total DRAM minus memory reserved for system use.">
+                            Unreserved
+                        </a>
+                    </div>
                 </div>
                 <div className="total-memory">
                     <div className="value">{total.value + ' ' + total.measure}</div>
-                    <div className="title">Total</div>
+                    <div className="title">
+                        <a data-toggle="tooltip" data-placement="bottom" data-trigger="hover"
+                            title="Total DRAM on the server.">
+                            Total
+                        </a>
+                    </div>
                 </div>
             </div>
         </div>;
-- 
2.21.0

